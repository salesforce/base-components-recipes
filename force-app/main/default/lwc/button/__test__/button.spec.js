import { createElement } from 'lwc';
import Element from 'c/button';

const createButton = (props = {}) => {
  const element = createElement('c-button', { is: Element });
  Object.assign(element, props);
  document.body.appendChild(element);
  return element;
};

describe('c-button', () => {
  it('should trigger focus when button is focused', () => {
    const element = createButton({});

    const evtListenerMock = jest.fn();
    element.addEventListener('focus', evtListenerMock);

    element.shadowRoot.querySelector('button').focus();
    expect(evtListenerMock.mock.calls).toHaveLength(1);
  });
  it('should trigger blur when button lost focus', () => {
    const element = createButton({});

    const evtListenerMock = jest.fn();
    element.addEventListener('blur', evtListenerMock);

    const btn = element.shadowRoot.querySelector('button');
    btn.focus();
    btn.blur();

    expect(evtListenerMock.mock.calls).toHaveLength(1);
  });
  it('should set title attribute on the button', () => {
    const element = createButton({ title: 'Click here' });

    const btn = element.shadowRoot.querySelector('button');

    expect(btn.getAttribute('title')).toBe('Click here');
    expect(element.getAttribute('title')).toBe(null);
  });
  it('should set aria-expanded attribute on the button when false', () => {
    const element = createButton({ ariaExpanded: 'false' });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(btn.getAttribute('aria-expanded')).toBe('false');
      expect(element.getAttribute('aria-expanded')).toBe(null);
    });
  });
  it('should set aria-expanded attribute on the button when true', () => {
    const element = createButton({ ariaExpanded: 'true' });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(btn.getAttribute('aria-expanded')).toBe('true');
      expect(element.getAttribute('aria-expanded')).toBe(null);
    });
  });
  it('shouldnt set aria-expanded attribute when boolean', () => {
    const element = createButton({ ariaExpanded: true });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(btn.getAttribute('aria-expanded')).toBe(null);
      expect(element.getAttribute('aria-expanded')).toBe(null);
    });
  });
  it('should set aria-live attribute on the button', () => {
    const element = createButton({ ariaLive: 'polite' });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(btn.getAttribute('aria-live')).toBe('polite');
      expect(element.getAttribute('aria-live')).toBe(null);
    });
  });
  it('should set aria-controls attribute on the button', () => {
    const element = createButton({ ariaControls: 'section1' });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(btn.getAttribute('aria-controls')).toEqual(
        expect.stringContaining('section1')
      );

      expect(element.getAttribute('aria-controls')).toBe(null);
    });
  });
  it('should set aria-label attribute on the button', () => {
    const element = createButton({ ariaLabel: 'btn1' });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(btn.getAttribute('aria-label')).toBe('btn1');
      expect(element.getAttribute('aria-label')).toBe(null);
    });
  });
  it('should set aria-describedby attribute on the button', () => {
    const element = createButton({ ariaDescribedBy: 'section1' });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(btn.getAttribute('aria-describedby')).toEqual(
        expect.stringContaining('section1')
      );

      expect(element.getAttribute('aria-describedby')).toBe(null);
    });
  });
  it('aria-attributes should be present and true', () => {
    const element = createButton({
      label: 'Test',
      title: 'Click here',
      ariaExpanded: 'true',
      ariaAtomic: 'true'
    });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(element).toMatchSnapshot();
      expect(btn.getAttribute('aria-expanded')).toBe('true');
      expect(btn.getAttribute('aria-atomic')).toBe('true');
    });
  });
  it('should set not set any aria-attributes', () => {
    const element = createButton({
      label: 'Test',
      title: 'Click here',
      tabIndex: '-1',
      ariaExpanded: null
    });

    const btn = element.shadowRoot.querySelector('button');

    return Promise.resolve().then(() => {
      expect(element).toMatchSnapshot();
      expect(btn.getAttribute('aria-expanded')).toBe(null);
      expect(btn.getAttribute('aria-atomic')).toBe(null);
    });
  });
  it('fires click event when button is clicked', () => {
    const element = createButton({ label: 'Submit' });
    const evtListenerMock = jest.fn();
    element.addEventListener('click', evtListenerMock);
    element.shadowRoot.querySelector('button').click();

    return Promise.resolve().then(() => {
      expect(evtListenerMock.mock.calls).toHaveLength(1);
    });
  });
});
