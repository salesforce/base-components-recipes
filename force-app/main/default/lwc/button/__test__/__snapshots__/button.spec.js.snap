// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lightning-button aria-attributes should be present and true 1`] = `
<lightning-button>
  #shadow-root(open)
    <button
      aria-atomic="true"
      aria-expanded="true"
      class="slds-button slds-button_neutral"
      title="Click here"
      type="button"
    >
      Test
    </button>
</lightning-button>
`;

exports[`lightning-button should set not set any aria-attributes 1`] = `
<lightning-button
  tabindex="-1"
>
  #shadow-root(open)
    <button
      class="slds-button slds-button_neutral"
      title="Click here"
      type="button"
    >
      Test
    </button>
</lightning-button>
`;
